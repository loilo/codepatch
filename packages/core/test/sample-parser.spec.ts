import { expect, it } from 'vitest'
import { SampleParser } from './fixtures/sample-parser'

// A bunch of snapshot tests checking the basic functionality of the sample parser

it('returns expected ASTs', () => {
  const parser = new SampleParser()

  expect(parser.run('(1,2,3)')).toMatchSnapshot()
  expect(parser.run('(1, 2, 3)')).toMatchSnapshot()
  expect(parser.run('(01, 20, 030)')).toMatchSnapshot()
  expect(parser.run('((1), 2, (3, (5, 6)), 7)')).toMatchSnapshot()
  expect(parser.run('(1, 2, ())')).toMatchSnapshot()
})

it('respects the "delimiter" option', () => {
  const parser = new SampleParser({ delimiter: '|' })

  expect(() => parser.run('(1,2,3)')).toThrow()
  expect(parser.run('(1|2|3)')).toMatchSnapshot()
  expect(parser.run('(1|(2|3)|4)')).toMatchSnapshot()
})

it('throws on invalid code', () => {
  const parser = new SampleParser()

  expect(() => parser.run('1,2,3')).toThrow()
  expect(() => parser.run('(1,2,3')).toThrow()
  expect(() => parser.run('(1/2)')).toThrow()
  expect(() => parser.run('(1 2)')).toThrow()
  expect(() => parser.run('(1), (2)')).toThrow()
})
