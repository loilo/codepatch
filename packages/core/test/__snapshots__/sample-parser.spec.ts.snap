// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`respects the "delimiter" option 1`] = `
{
  "children": [
    {
      "children": [
        {
          "loc": [
            1,
            2,
          ],
          "type": "integer",
          "value": 1,
        },
        {
          "loc": [
            3,
            4,
          ],
          "type": "integer",
          "value": 2,
        },
        {
          "loc": [
            5,
            6,
          ],
          "type": "integer",
          "value": 3,
        },
      ],
      "loc": [
        0,
        7,
      ],
      "type": "group",
    },
  ],
  "loc": [
    0,
    7,
  ],
  "type": "program",
}
`;

exports[`respects the "delimiter" option 2`] = `
{
  "children": [
    {
      "children": [
        {
          "loc": [
            1,
            2,
          ],
          "type": "integer",
          "value": 1,
        },
        {
          "children": [
            {
              "loc": [
                4,
                5,
              ],
              "type": "integer",
              "value": 2,
            },
            {
              "loc": [
                6,
                7,
              ],
              "type": "integer",
              "value": 3,
            },
          ],
          "loc": [
            3,
            8,
          ],
          "type": "group",
        },
        {
          "loc": [
            9,
            10,
          ],
          "type": "integer",
          "value": 4,
        },
      ],
      "loc": [
        0,
        11,
      ],
      "type": "group",
    },
  ],
  "loc": [
    0,
    11,
  ],
  "type": "program",
}
`;

exports[`returns expected ASTs 1`] = `
{
  "children": [
    {
      "children": [
        {
          "loc": [
            1,
            2,
          ],
          "type": "integer",
          "value": 1,
        },
        {
          "loc": [
            3,
            4,
          ],
          "type": "integer",
          "value": 2,
        },
        {
          "loc": [
            5,
            6,
          ],
          "type": "integer",
          "value": 3,
        },
      ],
      "loc": [
        0,
        7,
      ],
      "type": "group",
    },
  ],
  "loc": [
    0,
    7,
  ],
  "type": "program",
}
`;

exports[`returns expected ASTs 2`] = `
{
  "children": [
    {
      "children": [
        {
          "loc": [
            1,
            2,
          ],
          "type": "integer",
          "value": 1,
        },
        {
          "loc": [
            4,
            5,
          ],
          "type": "integer",
          "value": 2,
        },
        {
          "loc": [
            7,
            8,
          ],
          "type": "integer",
          "value": 3,
        },
      ],
      "loc": [
        0,
        9,
      ],
      "type": "group",
    },
  ],
  "loc": [
    0,
    9,
  ],
  "type": "program",
}
`;

exports[`returns expected ASTs 3`] = `
{
  "children": [
    {
      "children": [
        {
          "loc": [
            1,
            3,
          ],
          "type": "integer",
          "value": 1,
        },
        {
          "loc": [
            5,
            7,
          ],
          "type": "integer",
          "value": 20,
        },
        {
          "loc": [
            9,
            12,
          ],
          "type": "integer",
          "value": 30,
        },
      ],
      "loc": [
        0,
        13,
      ],
      "type": "group",
    },
  ],
  "loc": [
    0,
    13,
  ],
  "type": "program",
}
`;

exports[`returns expected ASTs 4`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "loc": [
                2,
                3,
              ],
              "type": "integer",
              "value": 1,
            },
          ],
          "loc": [
            1,
            4,
          ],
          "type": "group",
        },
        {
          "loc": [
            6,
            7,
          ],
          "type": "integer",
          "value": 2,
        },
        {
          "children": [
            {
              "loc": [
                10,
                11,
              ],
              "type": "integer",
              "value": 3,
            },
            {
              "children": [
                {
                  "loc": [
                    14,
                    15,
                  ],
                  "type": "integer",
                  "value": 5,
                },
                {
                  "loc": [
                    17,
                    18,
                  ],
                  "type": "integer",
                  "value": 6,
                },
              ],
              "loc": [
                13,
                19,
              ],
              "type": "group",
            },
          ],
          "loc": [
            9,
            20,
          ],
          "type": "group",
        },
        {
          "loc": [
            22,
            23,
          ],
          "type": "integer",
          "value": 7,
        },
      ],
      "loc": [
        0,
        24,
      ],
      "type": "group",
    },
  ],
  "loc": [
    0,
    24,
  ],
  "type": "program",
}
`;

exports[`returns expected ASTs 5`] = `
{
  "children": [
    {
      "children": [
        {
          "loc": [
            1,
            2,
          ],
          "type": "integer",
          "value": 1,
        },
        {
          "loc": [
            4,
            5,
          ],
          "type": "integer",
          "value": 2,
        },
        {
          "children": [],
          "loc": [
            7,
            9,
          ],
          "type": "group",
        },
      ],
      "loc": [
        0,
        10,
      ],
      "type": "group",
    },
  ],
  "loc": [
    0,
    10,
  ],
  "type": "program",
}
`;
